File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 8/13/2025, 8:24:28 PM
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack();
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if(stack.isEmpty() || ch == '(' || ch == '[' || ch == '{'){
                stack.push(ch);
            }else{
                if((ch == ')' && stack.peek() == '(') || (ch == ']' && stack.peek() == '[') || (ch == '}' && stack.peek() == '{')){
                 stack.pop();
                }
                else{
                return false;
            }
            }

        }
        if(stack.size()!=0){
            return false;
        }
        else{
            return true;
        }
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 8/13/2025, 8:24:22 PM
class Solution:
    def isValid(self, s: str) -> bool:
        myStack = []
        
        for i in s :
            if i=='(' or i=='{' or i=='[' or len(myStack) == 0 :
                myStack.append(i)
            elif i == ')' :
                if myStack[len(myStack)-1] == '(' :
                    myStack.pop()
                else :
                    return False
            elif i == '}' :
                if myStack[len(myStack)-1] == '{' :
                    myStack.pop()
                else :
                    return False
            elif i == ']' :
                if myStack[len(myStack)-1] == '[' :
                    myStack.pop()
                else :
                    return False
        if len(myStack) == 0 :
            return True
        return False
        